src/io/g15.rs

```diff
@@ -0,0 +1,202 @@
+    let idx = caps[1].parse::<usize>().unwrap();
+    let sid3 = caps[2].parse::<u64>().unwrap();
+    players[idx].sid3 = format!("[U:1:{}]", sid3);
+}
+
+// m_bValid[24] bool (true)     --match--> (player idx) (valid status (true/false))
+// I'm not sure what valid means - perhaps to with whether this is a current/in-game player idx
+// and to display this row in the scoreboard or not? Can someone look pls
+pub const REGEX_B_VALID: &str = r#"^m_bValid\[(\d+)\]\s+bool\s+\((false|true)\)$"#;
+pub fn parse_valid(caps: Captures, players: &mut Vec<G15Player>) {
+    let idx = caps[1].parse::<usize>().unwrap();
+    let valid_str = &caps[2];
+    let truth_value = matches!(valid_str, "true");
+    players[idx].valid = truth_value;
+}
+
+// m_iUserID[2] integer (68)    --match--> (player idx) (server id (as it appears in status))
+pub const REGEX_I_USERID: &str = r#"^m_iUserID\[(\d+)\]\s+integer\s+\((\d+)\)$"#;
+pub fn parse_userid(caps: Captures, players: &mut Vec<G15Player>) {
+    let idx = caps[1].parse::<usize>().unwrap();
+    let userid = caps[2].to_string();
+    players[idx].userid = userid;
+}
+
+// The G15Player struct contains all the data per player searched.
+#[derive(Debug, Clone)]
+pub struct G15Player {
+    name: String,     // eg "Lilith"
+    ping: u32,        // eg 21
+    score: u32,       // eg 16
+    deaths: u32,      // eg 5
+    sid3: String,     // eg [U:1:111216987]
+    team: u32,        // eg 3
+    health: u32,      // eg 125
+    ammo: u32,        // eg 6
+    connected: bool,  // eg true
+    valid: bool,      // eg true
+    alive: bool,      // eg true
+    userid: String,   // eg "301"
+}
+impl G15Player {
+    fn new() -> G15Player {
+        G15Player {
+            name: String::from("none"), 
+            ping: 0, 
+            score: 0, 
+            deaths: 0, 
+            sid3: String::from("none"), 
+            team: 0, 
+            health: 0, 
+            ammo: 0,
+            alive: false,
+            connected: false,
+            userid: String::from("0"),
+            valid: false,
+        }
+    }
+}
+
+/*
+  Parse a g15_dumpplayer string via Regex search.
+  We only extract useful data here, so drop most data.
+*/
+pub fn parse_g15(g15_log: String) -> Vec<G15Player> {
+    let mut players: Vec<G15Player> = vec![G15Player::new(); 34];
+    let patterns = [
+        RegMatch::new(REGEX_I_AMMO, parse_ammo), 
+        RegMatch::new(REGEX_SZ_NAME, parse_name),
+        RegMatch::new(REGEX_I_PING, parse_ping),
+        RegMatch::new(REGEX_I_SCORE, parse_score),
+        RegMatch::new(REGEX_I_DEATHS, parse_deaths),
+        RegMatch::new(REGEX_B_CONNECTED, parse_connected),
+        RegMatch::new(REGEX_I_TEAM, parse_team),
+        RegMatch::new(REGEX_B_ALIVE, parse_alive),
+        RegMatch::new(REGEX_I_HEALTH, parse_health),
+        RegMatch::new(REGEX_I_SID3, parse_sid3),
+        RegMatch::new(REGEX_B_VALID, parse_valid),
+        RegMatch::new(REGEX_I_USERID, parse_userid),
+        ];

```

Comment:
It would be nice to have this in a struct that we can build and reuse later (and since Regexes get compiled down to their automata and whatever, means we also don't have that extra latency of compiling 10 or so regexes every time the function is called).

I think it would be something like 
```rust
pub struct G15Parser {
    patterns: Vec<RegMatch>,
}

impl G15Parser {
    pub fn new() -> G15Parser { ... }
    pub fn parse_g15(&self, g15_log: &str) -> Vec<G15Player> { ... }
}
```

Reply:
should `g15_log` be `&str` or `String`?